crates/net/eth-wire/src/p2pstream.rs

```diff
@@ -519,14 +520,18 @@ pub fn set_capability_offsets(
     let mut shared_capability_names = BTreeSet::new();
 
     // find highest shared version of each shared capability
-    for capability in peer_capabilities {
+    for peer_capability in peer_capabilities {
         // if this is Some, we share this capability
-        if let Some(version) = our_capabilities_map.get(&capability.name) {
+        if our_capabilities.get(&peer_capability).is_some() {
             // If multiple versions are shared of the same (equal name) capability, the numerically
             // highest wins, others are ignored
-            if capability.version <= *version {
-                shared_capabilities.insert(capability.name.clone(), capability.version);
-                shared_capability_names.insert(capability.name);
+
+            let version = shared_capabilities.get(&peer_capability.name);
+            if version.is_none() ||
+                (version.is_some() && peer_capability.version > *version.unwrap())
```

Comment:
recommend using `expect` instead of `unwrap`, which signals why `None` is not possible
```suggestion
                (version.is_some() && peer_capability.version > *version.expect("is some; qed"))
```